Port WebExtension DeclarativeNetRequest Parsing

ItsJamie9494 opened the pull request:
#### cb007e60679aee6a8af66f54941d1727a618f147
<pre>
Port WebExtension DeclarativeNetRequest Parsing
<a href="https://webkit.org/b/283987">https://webkit.org/b/283987</a>

Reviewed by Timothy Hatcher.

Port DeclarativeNetRequest parsing to C++. This is the last portion of the manifest that must be ported to C++.

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::parseDeclarativeNetRequestRulesetDictionary): Deleted.
(WebKit::WebExtension::populateDeclarativeNetRequestPropertiesIfNeeded): Deleted.
(WebKit::WebExtension::declarativeNetRequestRulesets): Deleted.
(WebKit::WebExtension::declarativeNetRequestRuleset): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
(WebKit::WebExtension::parseDeclarativeNetRequestRulesetDictionary):
(WebKit::WebExtension::populateDeclarativeNetRequestPropertiesIfNeeded):
(WebKit::WebExtension::declarativeNetRequestRulesets):
(WebKit::WebExtension::declarativeNetRequestRuleset):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:

Canonical link: <a href="https://commits.webkit.org/287325@main">https://commits.webkit.org/287325@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/baed6530dfd46535f42af1d35dbca153ace78cae

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79164 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58193 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32537 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83790 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30355 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81297 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67289 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6461 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61936 "24 flakes 89 failures") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19849 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82231 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51982 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72142 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42241 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49333 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26270 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28729 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70450 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26694 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85183 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6468 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4478 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70179 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6631 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68014 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69427 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13478 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12290 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12226 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6423 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/12224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6366 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9828 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8158 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -29,6 +29,7 @@
 
 #if ENABLE(WK_WEB_EXTENSIONS)
 
+#include "WebExtensionConstants.h"
```
You should be able to remove this from `WebExtensionCocoa.mm` now too.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -2220,6 +2227,124 @@ void WebExtension::populateCommandsIfNeeded()
+
+    auto enabledBool = rulesetObject.getBoolean(declarativeNetRequestRuleEnabledManifestKey);
+    if (!enabledBool) {
+        error = createError(WebExtension::Error::InvalidDeclarativeNetRequest, "`declarative_net_request` missing enabled boolean"_s);
```
```suggestion
        error = createError(WebExtension::Error::InvalidDeclarativeNetRequest, WEB_UI_STRING("Missing or invalid `enabled` boolean for the `declarative_net_request` manifest entry.", "WKWebExtensionErrorInvalidDeclarativeNetRequestEntry description for missing enabled boolean"));
```

And re-run `update-webkit-localizable-strings` to update the strings from the source code.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -2220,6 +2227,124 @@ void WebExtension::populateCommandsIfNeeded()
+
+    auto jsonPath = rulesetObject.getString(declarativeNetRequestRulePathManifestKey);
+    if (jsonPath.isEmpty()) {
+        error = createError(WebExtension::Error::InvalidDeclarativeNetRequest, WEB_UI_STRING("Empty `declarative_net_request` JSON path.", "WKWebExtensionErrorInvalidDeclarativeNetRequestEntry description for empty JSON path"));
```
Ditto:
```suggestion
        error = createError(WebExtension::Error::InvalidDeclarativeNetRequest, WEB_UI_STRING("Empty or invalid `path` in `declarative_net_request` manifest entry.", "WKWebExtensionErrorInvalidDeclarativeNetRequestEntry description for empty or invalid path in declarative_net_request manifest entry"));
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -2220,6 +2227,124 @@ void WebExtension::populateCommandsIfNeeded()
+{
+    auto rulesetID = rulesetObject.getString(declarativeNetRequestRulesetIDManifestKey);
+    if (rulesetID.isEmpty()) {
+        error = createError(Error::InvalidDeclarativeNetRequest, WEB_UI_STRING("Empty `declarative_net_request` ruleset id.", "WKWebExtensionErrorInvalidDeclarativeNetRequestEntry description for empty ruleset id"));
```
The original should have been this way:
```suggestion
        error = createError(Error::InvalidDeclarativeNetRequest, WEB_UI_STRING("Empty or invalid `id` in `declarative_net_request` manifest entry.", "WKWebExtensionErrorInvalidDeclarativeNetRequestEntry description for empty or invalid id in declarative_net_request manifest entry"));
```

xeenon reviewed (changes requested):
Looks good. Some string nits.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -2220,6 +2227,124 @@ void WebExtension::populateCommandsIfNeeded()
+            continue;
+        }
+
+        auto ruleset = optionalRuleset.value();
```
```suggestion
        auto& ruleset = optionalRuleset.value();
```
or
```suggestion
        auto ruleset = WTFMove(optionalRuleset.value());
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -2220,6 +2227,124 @@ void WebExtension::populateCommandsIfNeeded()
+        seenRulesetIDs.add(ruleset.rulesetID);
+        ++rulesetCount;
+
+        m_declarativeNetRequestRulesets.append(ruleset);
```
```suggestion
        m_declarativeNetRequestRulesets.append(WTFMove(ruleset));
```

ItsJamie9494 committed:
Port WebExtension DeclarativeNetRequest Parsing
https://webkit.org/b/283987

Reviewed by Timothy Hatcher.

Port DeclarativeNetRequest parsing to C++. This is the last portion of the manifest that must be ported to C++.

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::parseDeclarativeNetRequestRulesetDictionary): Deleted.
(WebKit::WebExtension::populateDeclarativeNetRequestPropertiesIfNeeded): Deleted.
(WebKit::WebExtension::declarativeNetRequestRulesets): Deleted.
(WebKit::WebExtension::declarativeNetRequestRuleset): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
(WebKit::WebExtension::parseDeclarativeNetRequestRulesetDictionary):
(WebKit::WebExtension::populateDeclarativeNetRequestPropertiesIfNeeded):
(WebKit::WebExtension::declarativeNetRequestRulesets):
(WebKit::WebExtension::declarativeNetRequestRuleset):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:

Canonical link: https://commits.webkit.org/287325@main
